File: tools/ghidra/scripts/extract_functions.py
# Ghidra headless script: Extract all named functions with addresses and sizes
# Usage: Run via analyzeHeadless -postScript extract_functions.py <output_dir>
#
# Outputs:
#   functions_all.txt       - All named functions (addr, size, name)
#   functions_isp.txt       - ISP-related functions (hi_mpi_isp_*, secu_sensor_*, ot_mpi_*)
#   functions_cloud.txt     - Cloud/MQTT-related functions
#   functions_xuid.txt      - Internal IPC command handlers (HI_XUID_*)
#   functions_summary.txt   - Statistics summary

import os
import sys

args = getScriptArgs()
output_dir = args[0] if args else "."

program = getCurrentProgram()
fm = program.getFunctionManager()
listing = program.getListing()

print("[*] Extracting functions from: %s" % program.getName())

all_funcs = []
isp_funcs = []
cloud_funcs = []
xuid_funcs = []
ioctl_funcs = []
sensor_funcs = []
misc_interesting = []

isp_prefixes = ("hi_mpi_isp_", "ot_mpi_isp_", "hi_mpi_vi_", "ot_mpi_vi_",
                "hi_mpi_vpss_", "ot_mpi_vpss_", "hi_mpi_venc_", "ot_mpi_venc_",
                "hi_mpi_sys_", "ot_mpi_sys_")
sensor_prefixes = ("secu_sensor_",)
cloud_keywords = ("mqtt", "ali_", "iot_", "cloud", "thing_", "property",
                   "alink", "linkkit", "danale", "tutk", "p2p")
xuid_keywords = ("xuid", "HI_XUID", "hi_xuid")
ioctl_keywords = ("ioctl",)

func = fm.getFunctionAt(program.getMinAddress())
if func is None:
    func_iter = fm.getFunctions(True)
else:
    func_iter = fm.getFunctions(True)

count = 0
for func in func_iter:
    name = func.getName()
    addr = func.getEntryPoint()
    body = func.getBody()
    size = body.getNumAddresses()
    
    entry = "%s\t%d\t%s" % (addr, size, name)
    all_funcs.append(entry)
    count += 1
    
    name_lower = name.lower()
    
    # Categorize
    if any(name_lower.startswith(p) for p in isp_prefixes):
        isp_funcs.append(entry)
    
    if any(name_lower.startswith(p) for p in sensor_prefixes):
        sensor_funcs.append(entry)
    
    if any(kw in name_lower for kw in cloud_keywords):
        cloud_funcs.append(entry)
    
    if any(kw in name_lower for kw in xuid_keywords):
        xuid_funcs.append(entry)
    
    if any(kw in name_lower for kw in ioctl_keywords):
        ioctl_funcs.append(entry)
    
    # Other interesting functions
    for kw in ("reboot", "watchdog", "upgrade", "update", "factory", "reset",
               "record", "motion", "detect", "alarm", "night", "ircut",
               "flip", "mirror", "osd", "ptz", "motor", "audio", "speak",
               "brightness", "contrast", "saturation", "sharpness",
               "csc", "exposure", "white_balance", "gamma", "drc", "wdr",
               "encrypt", "decrypt", "password", "license", "key"):
        if kw in name_lower:
            misc_interesting.append(entry)
            break

print("[*] Total functions: %d" % count)
print("[*] ISP functions: %d" % len(isp_funcs))
print("[*] Sensor functions: %d" % len(sensor_funcs))
print("[*] Cloud functions: %d" % len(cloud_funcs))
print("[*] XUID functions: %d" % len(xuid_funcs))
print("[*] ioctl functions: %d" % len(ioctl_funcs))
print("[*] Other interesting: %d" % len(misc_interesting))

# Write output files
def write_list(filename, header, items):
    path = os.path.join(output_dir, filename)
    with open(path, "w") as f:
        f.write("# %s\n" % header)
        f.write("# Address\tSize\tName\n")
        f.write("# Extracted from: %s\n\n" % program.getName())
        for item in items:
            f.write(item + "\n")
    print("[*] Wrote %s (%d entries)" % (filename, len(items)))

write_list("functions_all.txt", "All named functions", all_funcs)
write_list("functions_isp.txt",
           "ISP/MPP functions (hi_mpi_isp_*, ot_mpi_*)",
           isp_funcs + sensor_funcs)
write_list("functions_cloud.txt",
           "Cloud/MQTT/IoT functions",
           cloud_funcs)
write_list("functions_xuid.txt",
           "Internal IPC (XUID) command handlers",
           xuid_funcs)
write_list("functions_ioctl.txt",
           "ioctl-related functions",
           ioctl_funcs)
write_list("functions_interesting.txt",
           "Other interesting functions (reboot, detect, record, etc.)",
           misc_interesting)

# Summary
summary_path = os.path.join(output_dir, "functions_summary.txt")
with open(summary_path, "w") as f:
    f.write("SUPERB BINARY FUNCTION ANALYSIS SUMMARY\n")
    f.write("=" * 60 + "\n\n")
    f.write("Binary: %s\n" % program.getName())
    f.write("Image base: %s\n" % program.getImageBase())
    f.write("Language: %s\n" % program.getLanguageID())
    f.write("Compiler: %s\n" % program.getCompilerSpec().getCompilerSpecID())
    f.write("\nFunction counts:\n")
    f.write("  Total named functions:  %d\n" % count)
    f.write("  ISP/MPP API:           %d\n" % len(isp_funcs))
    f.write("  Sensor control:        %d\n" % len(sensor_funcs))
    f.write("  Cloud/MQTT/IoT:        %d\n" % len(cloud_funcs))
    f.write("  XUID IPC handlers:     %d\n" % len(xuid_funcs))
    f.write("  ioctl-related:         %d\n" % len(ioctl_funcs))
    f.write("  Other interesting:     %d\n" % len(misc_interesting))

print("[*] Summary written to functions_summary.txt")
print("[*] extract_functions.py complete")

File: tools/ghidra/scripts/ExtractFunctions.java
// Ghidra headless script: Extract all named functions with addresses and sizes
// Run via: analyzeHeadless ... -postScript ExtractFunctions.java <output_dir>
//@category Analysis

import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.*;
import ghidra.program.model.address.*;
import java.io.*;
import java.util.*;

public class ExtractFunctions extends GhidraScript {

    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        String outputDir = args.length > 0 ? args[0] : ".";

        FunctionManager fm = currentProgram.getFunctionManager();
        int total = 0;
        List<String> allFuncs = new ArrayList<>();
        List<String> ispFuncs = new ArrayList<>();
        List<String> sensorFuncs = new ArrayList<>();
        List<String> cloudFuncs = new ArrayList<>();
        List<String> xuidFuncs = new ArrayList<>();
        List<String> ioctlFuncs = new ArrayList<>();
        List<String> interestingFuncs = new ArrayList<>();

        String[] ispPrefixes = {"hi_mpi_isp_", "ot_mpi_isp_", "hi_mpi_vi_", "ot_mpi_vi_",
            "hi_mpi_vpss_", "ot_mpi_vpss_", "hi_mpi_venc_", "ot_mpi_venc_",
            "hi_mpi_sys_", "ot_mpi_sys_", "hi_mpi_ai_", "ot_mpi_ai_",
            "hi_mpi_ao_", "ot_mpi_ao_", "hi_mpi_aenc_", "ot_mpi_aenc_",
            "hi_mpi_adec_", "ot_mpi_adec_"};
        String[] sensorPrefixes = {"secu_sensor_"};
        String[] cloudKeywords = {"mqtt", "ali_", "iot_", "cloud", "thing_", "property",
            "alink", "linkkit", "danale", "tutk", "p2p"};
        String[] xuidKeywords = {"xuid", "hi_xuid"};
        String[] interestingKeywords = {"reboot", "watchdog", "upgrade", "update", "factory",
            "reset", "record", "motion", "detect", "alarm", "night", "ircut",
            "flip", "mirror", "osd", "ptz", "motor", "audio", "speak",
            "brightness", "contrast", "saturation", "sharpness",
            "csc", "exposure", "white_balance", "gamma", "drc", "wdr",
            "encrypt", "decrypt", "password", "license"};

        FunctionIterator funcIter = fm.getFunctions(true);
        while (funcIter.hasNext() && !monitor.isCancelled()) {
            Function func = funcIter.next();
            String name = func.getName();
            String addr = func.getEntryPoint().toString();
            long size = func.getBody().getNumAddresses();
            String entry = addr + "\t" + size + "\t" + name;
            String nameLower = name.toLowerCase();

            allFuncs.add(entry);
            total++;

            for (String p : ispPrefixes) {
                if (nameLower.startsWith(p)) { ispFuncs.add(entry); break; }
            }
            for (String p : sensorPrefixes) {
                if (nameLower.startsWith(p)) { sensorFuncs.add(entry); break; }
            }
            boolean isCloud = false;
            for (String kw : cloudKeywords) {
                if (nameLower.contains(kw)) { cloudFuncs.add(entry); isCloud = true; break; }
            }
            for (String kw : xuidKeywords) {
                if (nameLower.contains(kw)) { xuidFuncs.add(entry); break; }
            }
            if (nameLower.contains("ioctl")) {
                ioctlFuncs.add(entry);
            }
            for (String kw : interestingKeywords) {
                if (nameLower.contains(kw)) { interestingFuncs.add(entry); break; }
            }
        }

        println("[*] Total functions: " + total);
        println("[*] ISP: " + ispFuncs.size() + ", Sensor: " + sensorFuncs.size() +
                ", Cloud: " + cloudFuncs.size() + ", XUID: " + xuidFuncs.size() +
                ", ioctl: " + ioctlFuncs.size() + ", Interesting: " + interestingFuncs.size());

        writeList(outputDir, "functions_all.txt", "All named functions", allFuncs);
        writeList(outputDir, "functions_isp.txt", "ISP/MPP API functions", ispFuncs);
        writeList(outputDir, "functions_sensor.txt", "Sensor control (secu_sensor_*)", sensorFuncs);
        writeList(outputDir, "functions_cloud.txt", "Cloud/MQTT/IoT functions", cloudFuncs);
        writeList(outputDir, "functions_xuid.txt", "Internal IPC (XUID) handlers", xuidFuncs);
        writeList(outputDir, "functions_ioctl.txt", "ioctl-related functions", ioctlFuncs);
        writeList(outputDir, "functions_interesting.txt", "Other interesting functions", interestingFuncs);

        // Summary
        PrintWriter pw = new PrintWriter(new FileWriter(new File(outputDir, "functions_summary.txt")));
        pw.println("SUPERB BINARY FUNCTION ANALYSIS SUMMARY");
        pw.println("=".repeat(60));
        pw.println();
        pw.println("Binary: " + currentProgram.getName());
        pw.println("Image base: " + currentProgram.getImageBase());
        pw.println("Language: " + currentProgram.getLanguageID());
        pw.println("Compiler: " + currentProgram.getCompilerSpec().getCompilerSpecID());
        pw.println();
        pw.println("Function counts:");
        pw.println("  Total named:     " + total);
        pw.println("  ISP/MPP API:     " + ispFuncs.size());
        pw.println("  Sensor control:  " + sensorFuncs.size());
        pw.println("  Cloud/MQTT/IoT:  " + cloudFuncs.size());
        pw.println("  XUID IPC:        " + xuidFuncs.size());
        pw.println("  ioctl-related:   " + ioctlFuncs.size());
        pw.println("  Interesting:     " + interestingFuncs.size());
        pw.close();

        println("[*] ExtractFunctions complete");
    }

    private void writeList(String dir, String filename, String header, List<String> items) throws IOException {
        PrintWriter pw = new PrintWriter(new FileWriter(new File(dir, filename)));
        pw.println("# " + header);
        pw.println("# Address\tSize\tName");
        pw.println("# Extracted from: " + currentProgram.getName());
        pw.println();
        for (String item : items) {
            pw.println(item);
        }
        pw.close();
        println("[*] Wrote " + filename + " (" + items.size() + " entries)");
    }
}
